romankondratev-BJS2-61678: Supercow!

romankondratev opened the pull request.

romankondratev committed:
BJS2-61678 feat: реализовать игровую механику с игроками и боссом

- Добавлен класс `Player` с логикой участия в сражении.
- Добавлен класс `Boss` с синхронизацией входа/выхода игроков из битвы.
- Добавлен класс `Main` для запуска игры, управления потоками и обработки игроков.
- Реализована многопоточная логика с ожиданием и уведомлением игроков.
- Поддержка ограничения количества активных участников битвы.

RuslanZinovyev commented on src/main/java/school/faang/task_61678/Boss.java:
```diff
@@ -0,0 +1,52 @@
+    private final Object lock = new Object();
+    private final int maxPlayers;
+
+    private int currentPlayers;
```
Надо придерживаться общего стиля форматирования, отступ ни к чему тут.

romankondratev replied:
А какой общий стиль форматирования?
Я отделил `private final` от `private`

RuslanZinovyev commented on src/main/java/school/faang/task_61678/Player.java:
```diff
@@ -0,0 +1,30 @@
+        log.info("Player: {}, Status: Battle Started", name);
+
+        performBattle();
+        boss.leaveBattle(this);
```
этот метод никогда не исполнится если joinBattle выбросит исключение, я бы не увлекался `SneakyThrows` и сделал `try catch finally` тут

romankondratev committed:
BJS2-61678 fix: добавить обработку исключений в процессе битвы игрока

- Обернул логику `doBattle` в блок try-catch для обработки исключений
- Добавил логирование ошибок при невозможности игрока присоединиться к битве
- Обеспечил вызов `leaveBattle` в блоке `finally`
